File: Spring/Chapter4/java_demo/demo/src/main/java/com/proxyclass/ProxyClassAspect.java
package com.proxyclass;

import org.aopalliance.intercept.MethodInterceptor;
import org.aopalliance.intercept.MethodInvocation;

public class ProxyClassAspect implements MethodInterceptor {
    @Override
    public Object invoke(MethodInvocation methodInvocation) throws Throwable {
        check();
        expect();
        Object obj = methodInvocation.proceed();
        log();
        monitor();
        return obj;
    }

    private void check()
    {
        System.out.println("模拟权限控制");
    }

    private void expect()
    {
        System.out.println("模拟异常处理");
    }

    private void log()
    {
        System.out.println("模拟日志记录");
    }

    private void monitor()
    {
        System.out.println("性能监测");
    }
}

File: Spring/Chapter4/kotlin_demo/demo/src/main/kotlin/proxyclass/ProxyClassAspect.kt
package proxyclass

import org.aopalliance.intercept.MethodInterceptor
import org.aopalliance.intercept.MethodInvocation

class ProxyClassAspect:MethodInterceptor{
    override fun invoke(invocation: MethodInvocation?): Any? {
        check()
        expect()
        val obj = invocation?.proceed()
        log()
        monitor()
        return obj
    }

    private fun check() = println("模拟权限控制")

    private fun expect() = println("模拟异常处理")

    private fun log() = println("模拟日志记录")

    private fun monitor() = println("性能监测")
}